### PR TITLE
KAFKA-7021: Reuse source based on config

### PR DESCRIPTION
**Pull request by guozhangwang**
This PR actually contains two changes:

1) leverage on the TOPOLOGY_OPTIMIZATION config to "adjust" the topology internally to reuse the source topic.

2) fixed a long dangling bug that whenever source topic is reused as changelog topic, write the checkpoint file for the consumed offset, this is done by union the `ackedOffset` from the producer, plus the `consumed` offset from the consumer, note we will priori ackedOffset since the same topic may show up in both (think about repartition topic), by doing this the consumed offset from source topics can be treated as checkpointed offset when reuse happens.

3) added a few unit and integration tests with / wo  the reusing, and make sure the restoration, standby task, and internal topic creation behaviors are all correct.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by guozhangwang** [comment]
@abbccdda @Ishiihara 

@vvcephei @mjsax @bbejeck

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java [hunk 1]
```diff
@@ -167,7 +167,7 @@ public String toString(final String indent) {
         return sb.toString();
     }
 
-    protected Map<TopicPartition, Long> recordCollectorOffsets() {
+    protected Map<TopicPartition, Long> activeTaskCheckpointableOffsets() {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java:170]: nit renaming.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 2]
```diff
@@ -391,8 +391,13 @@ void commit(final boolean startNewTransaction) {
     }
 
     @Override
-    protected Map<TopicPartition, Long> recordCollectorOffsets() {
-        return recordCollector.offsets();
+    protected Map<TopicPartition, Long> activeTaskCheckpointableOffsets() {
+        final Map<TopicPartition, Long> checkpointableOffsets = recordCollector.offsets();
+        for (Map.Entry<TopicPartition, Long> entry : consumedOffsets.entrySet()) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: This is the actual fix.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 3]
```diff
@@ -212,6 +212,11 @@
      */
     public static final String OPTIMIZE = "all";
 
+    /**
+     * Config value for parameter (@link #TOPOLOGY_OPTIMIZATION "topology.optimization" for enabling topology optimization
+     */
+    public static final String OPTIMIZE_AT_20 = "2.0";
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: Why do we need this? Should we not use `"all"` ? Also, this would be  public API change -- we can still update the corresponding KIP, but I don't think we need this, do we?

File: streams/src/main/java/org/apache/kafka/streams/Topology.java [hunk 4]
```diff
@@ -777,4 +778,23 @@ public synchronized TopologyDescription describe() {
         return internalTopologyBuilder.describe();
     }
 
+    // Adjust the generated topology based on the configs.
+    // Not exposed as public API and should be removed post 2.0
+    Topology adjust(final StreamsConfig config) {
+        final boolean enableOptimization20 = config.getString(StreamsConfig.TOPOLOGY_OPTIMIZATION).equals(StreamsConfig.OPTIMIZE_AT_20);
+
+        if (enableOptimization20) {
+            for (Map.Entry<StoreBuilder, String> entry : internalTopologyBuilder.storeToSourceChangelogTopic.entrySet()) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java]: nit: add `final`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java [hunk 5]
```diff
@@ -566,16 +575,22 @@ public final void connectProcessorAndStateStores(final String processorName,
...
     public final void connectSourceStoreAndTopic(final String sourceStoreName,
-                                                  final String topic) {
+                                                 final String topic) {
         if (storeToChangelogTopic.containsKey(sourceStoreName)) {
             throw new TopologyException("Source store " + sourceStoreName + " is already added.");
         }
         storeToChangelogTopic.put(sourceStoreName, topic);
     }
 
-    // TODO: this method is only used by DSL and we might want to refactor this part
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:578]: Why did you remove those comments? I still think, we should refactor some parts here (also applies to other TODOs below).

File: streams/src/test/java/org/apache/kafka/streams/integration/RestoreIntegrationTest.java [hunk 6]
```diff
@@ -112,24 +121,106 @@ public void shutdown() {
     }
 
     @Test
-    public void shouldRestoreState() throws ExecutionException, InterruptedException {
+    public void shouldRestoreStateFromSourceTopic() throws InterruptedException, IOException {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/RestoreIntegrationTest.java]: nit: simplify `InterruptedException, IOException` to `Exception`

File: streams/src/test/java/org/apache/kafka/streams/integration/TableTableJoinIntegrationTest.java [hunk 7]
```diff
@@ -55,8 +55,8 @@ public void prepareTopology() throws InterruptedException {
         appID = "table-table-join-integration-test";
 
         builder = new StreamsBuilder();
-        leftTable = builder.table(INPUT_TOPIC_LEFT);
-        rightTable = builder.table(INPUT_TOPIC_RIGHT);
+        leftTable = builder.table(INPUT_TOPIC_LEFT, Materialized.<Long, String, KeyValueStore<Bytes, byte[]>>as("left").withLoggingDisabled());
+        rightTable = builder.table(INPUT_TOPIC_RIGHT, Materialized.<Long, String, KeyValueStore<Bytes, byte[]>>as("right").withLoggingDisabled());
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/TableTableJoinIntegrationTest.java:59]: Do we need to disable logging explicitly here?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java [hunk 8]
```diff
@@ -72,11 +72,7 @@ public InternalStreamsBuilder(final InternalTopologyBuilder internalTopologyBuil
     public <K, V> KTable<K, V> table(final String topic,
                                      final ConsumedInternal<K, V> consumed,
                                      final MaterializedInternal<K, V, KeyValueStore<Bytes, byte[]>> materialized) {
-        // explicitly disable logging for source table materialized stores
-        materialized.withLoggingDisabled();
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:76]: Why do we remove this?

File: streams/src/main/java/org/apache/kafka/streams/Topology.java [hunk 9]
```diff
@@ -777,4 +778,23 @@ public synchronized TopologyDescription describe() {
         return internalTopologyBuilder.describe();
     }
 
+    // Adjust the generated topology based on the configs.
+    // Not exposed as public API and should be removed post 2.0
+    Topology adjust(final StreamsConfig config) {
+        final boolean enableOptimization20 = config.getString(StreamsConfig.TOPOLOGY_OPTIMIZATION).equals(StreamsConfig.OPTIMIZE_AT_20);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java]: revert the check to avoid NPE: `StreamsConfig.OPTIMIZE_AT_20.equals(config.getString(StreamsConfig.TOPOLOGY_OPTIMIZATION))`

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 10]
```diff
@@ -535,7 +535,7 @@ public void onRestoreEnd(final TopicPartition topicPartition, final String store
      */
     public KafkaStreams(final Topology topology,
                         final Properties props) {
-        this(topology.internalTopologyBuilder, new StreamsConfig(props), new DefaultKafkaClientSupplier());
+        this(topology.adjust(new StreamsConfig(props)).internalTopologyBuilder, new StreamsConfig(props), new DefaultKafkaClientSupplier());
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: why do we call `adjust` in each constructor? Doing it once in the private constructor should be sufficient?

File: streams/src/main/java/org/apache/kafka/streams/Topology.java [hunk 11]
```diff
@@ -777,4 +778,23 @@ public synchronized TopologyDescription describe() {
...
+
+        if (enableOptimization20) {
+            for (Map.Entry<StoreBuilder, String> entry : internalTopologyBuilder.storeToSourceChangelogTopic.entrySet()) {
+                final StoreBuilder storeBuilder = entry.getKey();
+                final String topicName = entry.getValue();
+
+                // update store map to disable logging for this store
+                storeBuilder.withLoggingDisabled();
+                internalTopologyBuilder.addStateStore(storeBuilder, true);
+                internalTopologyBuilder.connectSourceStoreAndTopic(storeBuilder.name(), topicName);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java]: Just for my understanding: we are overwriting the changelog-topic name with the source-topic name here?

File: streams/src/main/java/org/apache/kafka/streams/Topology.java [hunk 12]
```diff
@@ -777,4 +778,23 @@ public synchronized TopologyDescription describe() {
...
+        final boolean enableOptimization20 = config.getString(StreamsConfig.TOPOLOGY_OPTIMIZATION).equals(StreamsConfig.OPTIMIZE_AT_20);
+
+        if (enableOptimization20) {
+            for (Map.Entry<StoreBuilder, String> entry : internalTopologyBuilder.storeToSourceChangelogTopic.entrySet()) {
+                final StoreBuilder storeBuilder = entry.getKey();
+                final String topicName = entry.getValue();
+
+                // update store map to disable logging for this store
+                storeBuilder.withLoggingDisabled();
+                internalTopologyBuilder.addStateStore(storeBuilder, true);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java]: why do we need to add the store again? Isn't `storeBuilder` the same object that is already present and we overwrite it with itself here? `storeBuilder.withLoggingDisabled()` is a mutable call, right?

**Review by mjsax — COMMENTED** [review]
Couple of comments/question.

I think I understand to overall idea and it makes sense -- need think about some details a little more to really fully understand the change.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java [hunk 13]
```diff
@@ -121,6 +121,9 @@
 
     private Map<Integer, Set<String>> nodeGroups = null;
 
+    // this is only temporary for 2.0 and should be removed
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java]: Maybe mark it TODO so we can search for it later.

File: streams/src/test/java/org/apache/kafka/streams/integration/RestoreIntegrationTest.java (hunk 6 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/RestoreIntegrationTest.java]: I disagree. It can be nicer for the reader to know which checked exceptions we're potentially going to throw.

But it doesn't matter much, so do what you want ;)

File: streams/src/main/java/org/apache/kafka/streams/Topology.java (hunk 9 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/Topology.java]: I'm not sure I'm reading this right... If I say `optimize:=all`, we won't do this optimization, right? It seems like you want to check for `2.0` *or* `all`

**Comment by vvcephei** [comment]
I left a few comments, and I was also curious about several of @mjsax 's concerns, but assuming you hash those out, I'm +1 overall.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 10 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: I could be missing something, but the private constructor takes an `InternalTopologyBuilder` instance and `adjust` needs to execute to enable source topic reuse before returning the `InternalTopologyBuilder`.  
But maybe we could shift the `adjust` method to the `InternalTopologyBuilder` thus only need to call once in the private constructor.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java (hunk 3 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: I agree.  Since this our first optimization, `"all"` should be sufficient.  But this raises another good point if we plan to track optimization versions are we setting ourselves up for a KIP for each optimization release?

File: streams/src/test/java/org/apache/kafka/streams/integration/RestoreIntegrationTest.java (hunk 6 above)
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/integration/RestoreIntegrationTest.java]: For API calls I agree, but for testing, I think to collapse to `Exception` is ok.

**Review by bbejeck — COMMENTED** [review]
Overall I agree with the approach.  Just left a couple of minor comments.  I'll probably take another pass over the tests later to confirm my understanding of the offset restore fix

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java (hunk 3 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: 1. Sounds good. I'm reverting to use `all`.
2. Moving forward, not needing one KIP for each release. We can use a single KIP adding the mechanism, and if two versions are compatible (say if no new incompatible optimizations introduced from X to Y), we can just document it in upgrade section.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java (hunk 5 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:578]: After reviewed Bill's part II PR for optimization, which enhanced the InternalStreamsBuilder for logical plan generation, I think it makes less sense to refactor these into it.

File: streams/src/test/java/org/apache/kafka/streams/integration/TableTableJoinIntegrationTest.java (hunk 7 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/integration/TableTableJoinIntegrationTest.java:59]: Yes. We want to not restore the stores in between tests (it is tricker to delete the changelogs since appId changes for each case, so I think this is easier).

File: streams/src/main/java/org/apache/kafka/streams/Topology.java (hunk 9 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/Topology.java]: ack

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java (hunk 8 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:76]: The actual fix works as following: in the parsing phase, we create the materialized store normally, which means not enforcing disable logging. And then in `adjust` we will modify the store, to 1) disable logging, 2) change its changelog topic to the source topic.

As I mentioned in another HOTFIX PR, note that the term "changelog" topic now has two semantics meaning: 1) for restoring, 2) for appending updates to. For example with optimizations turned on, the changelog for 1) will be the source topic, and for 2) it will be none. There are some tricky separate code paths handling those cases now.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 10 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: What I originally thought is exactly as @bbejeck explained. I think changing to `InternalTopologyBuilder` would be a better idea, will try it out.

File: streams/src/main/java/org/apache/kafka/streams/Topology.java (hunk 11 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/Topology.java]: Yes.

File: streams/src/main/java/org/apache/kafka/streams/Topology.java (hunk 12 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/Topology.java]: Unfortunately not.. note that `addStateStore` directly constructs the `StoreFactory` and put that into the map `stateFactories`. And `StoreFactory` is immutable. So I have to overwrite the key with a new StoreFactory in this way.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java (hunk 13 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java]: acj,

File: streams/src/test/java/org/apache/kafka/streams/integration/RestoreIntegrationTest.java (hunk 6 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/RestoreIntegrationTest.java]: For regular API calls (ie, "main code") we should list them. For a test, no exception should ever be thrown, and if, the test fails. Which exception is irrelevant IMHO.

File: streams/src/test/java/org/apache/kafka/streams/integration/TableTableJoinIntegrationTest.java (hunk 7 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/TableTableJoinIntegrationTest.java:59]: Ack.

File: streams/src/main/java/org/apache/kafka/streams/Topology.java (hunk 12 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/Topology.java]: Ack.

**Comment by mjsax** [comment]
@guozhangwang Build failed with checkstyle error

**Comment by guozhangwang** [comment]
Found the issue with failed system test, filed another PR: https://github.com/apache/kafka/pull/5170

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java (hunk 2 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: nit: add `final`

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java [hunk 14]
```diff
@@ -273,18 +273,39 @@ public void shouldUseDefaultNodeAndStoreNames() {
...
+    public void shouldNotReuseSourceTopicAsChangelogsByDefault() {
+        final String topic = "topic";
+        builder.table(topic, Materialized.<Long, String, KeyValueStore<Bytes, byte[]>>as("store"));
+
+        final InternalTopologyBuilder internalTopologyBuilder = TopologyWrapper.getInternalTopologyBuilder(builder.build());
+        internalTopologyBuilder.setApplicationId("appId");
+
+        assertThat(internalTopologyBuilder.build().storeToChangelogTopic(), equalTo(Collections.singletonMap("store", "appId-store-changelog")));
+
+        assertThat(internalTopologyBuilder.getStateStores().keySet(), equalTo(Collections.singleton("store")));
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:305]: Should we add this check to `shouldReuseSourceTopicAsChangelogsWithOptimization20`, too?

File: streams/src/test/java/org/apache/kafka/streams/integration/RestoreIntegrationTest.java [hunk 15]
```diff
@@ -112,24 +120,106 @@ public void shutdown() {
...
 
-        createStateForRestoration();
+        final Properties props = props(APPID);
+        props.put(StreamsConfig.TOPOLOGY_OPTIMIZATION, StreamsConfig.OPTIMIZE);
+
+        // restoring from 1000 to 4000 (committed), and then process from 4000 to 5000 on each of the two partitions
+        final int offsetLimitDelta = 1000;
+        final int offsetCheckpointed = 1000;
+        createStateForRestoration(INPUT_STREAM);
+        setCommittedOffset(INPUT_STREAM, offsetLimitDelta);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/RestoreIntegrationTest.java:131]: Why do we use a delta, here, instead of the actual offset we want to commit (ie, 4000) ? Might be simpler to understand the test?

File: streams/src/test/java/org/apache/kafka/streams/integration/RestoreIntegrationTest.java [hunk 16]
```diff
@@ -112,24 +120,106 @@ public void shutdown() {
...
+                .write(Collections.singletonMap(new TopicPartition(APPID + "-store-changelog", 0), (long) offsetCheckpointed));
+        new OffsetCheckpoint(new File(stateDirectory.directoryForTask(new TaskId(0, 1)), ".checkpoint"))
+                .write(Collections.singletonMap(new TopicPartition(APPID + "-store-changelog", 1), (long) offsetCheckpointed));
 
         final CountDownLatch startupLatch = new CountDownLatch(1);
-        kafkaStreams = new KafkaStreams(builder.build(), props(applicationId));
+        final CountDownLatch shutdownLatch = new CountDownLatch(1);
+
+        builder.table(INPUT_STREAM, Consumed.with(Serdes.Integer(), Serdes.Integer()), Materialized.as("store"))
+                .toStream()
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/RestoreIntegrationTest.java:210]: nit: indention should only be 4 spaces?

File: streams/src/test/java/org/apache/kafka/streams/integration/RestoreIntegrationTest.java [hunk 17]
```diff
@@ -279,39 +371,41 @@ public void close() {
...
         consumerConfig.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, IntegerDeserializer.class);
 
         final Consumer consumer = new KafkaConsumer(consumerConfig);
-        final List<TopicPartition> partitions = Arrays.asList(new TopicPartition(INPUT_STREAM, 0),
-                                                              new TopicPartition(INPUT_STREAM, 1));
+        final List<TopicPartition> partitions = Arrays.asList(new TopicPartition(topic, 0),
+                new TopicPartition(topic, 1));
 
         consumer.assign(partitions);
         consumer.seekToEnd(partitions);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/RestoreIntegrationTest.java]: If we rewrite to use "absolute position" instead of delta, we can remove this.

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java (hunk 14 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:305]: ack.

File: streams/src/test/java/org/apache/kafka/streams/integration/RestoreIntegrationTest.java (hunk 15 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/integration/RestoreIntegrationTest.java:131]: The reason is that in line 150, I need to check exactly how many data were processed, which should be the diff between committed offset to log end offset.

However, for log end offset it is not always numKeys / 2 (num.partitions). In my tests I saw for example 5005 and 4995. So I have to use the limit to say `commit at the log end offset minus that delta`.

**Review by vvcephei — APPROVED** [review]

File: streams/src/test/java/org/apache/kafka/streams/integration/RestoreIntegrationTest.java (hunk 15 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/RestoreIntegrationTest.java:131]: Ack.

**Review by mjsax — APPROVED** [review]
LGTM. Thanks for the patch.

**Review by bbejeck — APPROVED** [review]

**Comment by guozhangwang** [comment]
Thanks for the reviews, I've merged it to trunk.

Will provide a separate PR for upgrade docs, and have separate PRs for older branches.

**Comment by guozhangwang** [comment]
Cherry-picked to 2.0
